fix(server): Properly build ML predict URL

safehome-jdev opened the pull request:
#9676 - A trailing slash at the end of the machine learning URL breaks the requests to the ML endpoint.

- Created a new enum to filter for URL specific types
- Added a `validationMessage` variable to provide information to the front-end for user feedback
- Added an `isInvalid` variable to provide a state for the tailwind classes
- These changes do not prevent a user from pressing save

validation error with message:
<img width="704" alt="image" src="https://github.com/immich-app/immich/assets/65427171/1268337e-f778-4097-90c0-70dce1b485ec">

valid input:
<img width="704" alt="image" src="https://github.com/immich-app/immich/assets/65427171/629c8a0a-20a2-40de-b008-ef2f9ee562a0">

bo0tzz reviewed (commented):
Shouldn't this be serverside validation instead?

Furthermore, validating URLs with complex regex is not great. It'd be much better to use something like [URL.canParse()](https://developer.mozilla.org/en-US/docs/Web/API/URL/canParse_static) instead, or an actual validation library.

martabal commented:
yep, IMO that's should be a server-side validation. Technically, `http://immich-machine-learning:3000/` is a valid URL and it would be misleading to not accept it especially when the input title and the description are `URL` and `URL of the machine learning server`. 

But having a way to alert/block when a URL is not a valid URL like in your first example would be amazing.

safehome-jdev commented:
@bo0tzz @martabal making this server-side would mean validation at both locations to the same input, no? I say that because in order to provide real-time user feedback, we need to manipulate the DOM; otherwise we'd need to provide an API response that their input is incorrect after pressing save and highlight the objects that are incorrect with some form of object that identifies the form inputs.

Also I didn't know that `urlParse()` accepted non-Latin characters. I'll have to give that a test

bo0tzz commented:
I just tested and Immich already rejects invalid machine-learning URLs at the API level, with a clear error message in the UI, so none of these clientside changes are necessary. What is still necessary, and what is really the problem in #9676, is to properly build the URL when calling the machine learning service in a way that does not break if there is a trailing slash. That would be done by building a URL object, rather than the current string concatenation:
```javascript
let baseUrl = "http://immich-machine-learning:3003/";
let url = new URL("/predict", baseUrl);
// use url in fetch call
```

safehome-jdev committed:
New URL via URL constructor and not string concatenation

safehome-jdev commented:
Removed previous UI validation suggestions and input highlighting due to API validation performing this. Modified server-side code to use the `new URL` constructor instead of string con-🐈.

bo0tzz reviewed (approved):
Thanks!
